Telemetry changes

JamesWTruher opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

- Stop sending telemetry for application type
- Send telemetry for parameters used when running pwsh
- Include version number when reporting module import

<!-- Summarize your PR between here and the checklist. -->

## PR Context

Our telemetry ingestion is currently very high with between 75-80% being taken up by ApplicationType metrics. The data doesn't seem to be showing a lot of variation over time, and isn't providing the value it did initially for its cost. We also have questions which are difficult to answer with the current telemetry (primarily disambiguating automation from interactive sessions). This PR includes metrics of how parameters of pwsh.exe are used. To reduce the payload size, the parameters are created as a bitmap which can then be investigated via kusto queries. In addition to collecting the used parameters, we are now capturing the value of ExecutionPolicy to help us understand the distribution of how this is being used. Lastly, this PR will also collect the version number of the loaded module where we can report it. With this we can make determinations as to how new modules are picked up, and how long older modules remain in use.


<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

JamesWTruher committed:
Stop sending telemetry for the type of entity being executed.

Also add telemetry for the parameters provided to pwsh.

JamesWTruher committed:
Remove the telemetry send from the pipeline.

But keep support for it in the telemetry client.

JamesWTruher committed:
change parameter bitmap to something more sensible.

Tidy up alignment.

JamesWTruher committed:
Turn off developer mode for telemetry send.

Update comments to improve understanding.

JamesWTruher committed:
Do not send the start up event out of the send metric method.

This is not needed and can result in us not sending the telemetry that includes the parameter data.

JamesWTruher committed:
first round of CodeFactor issues.

JamesWTruher committed:
Use a default value for the node if we can't create the telemetry.uuid file.

This handles the case where we have a read-only file system and want to continue to capture telemetry.

JamesWTruher committed:
Fix test to reflect new behavior in startup telemetry send.

StevenBucher98 commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -642,6 +649,28 @@ private static bool GetEnvironmentVariableAsBool(string name, bool defaultValue)
+        /// <summary>
+        /// Send module load telemetry as a metric.
+        /// For modules we send the module name (if allowed), and the version.
+        /// Some modules (CIM) will continue use the string alternative method.
```
I am unsure what CIM is here

JamesWTruher replied:
discussed here: https://docs.microsoft.com/en-us/powershell/module/cimcmdlets/?view=powershell-7.2

StevenBucher98 commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -642,6 +649,28 @@ private static bool GetEnvironmentVariableAsBool(string name, bool defaultValue)
+        /// <summary>
+        /// Send module load telemetry as a metric.
+        /// For modules we send the module name (if allowed), and the version.
+        /// Some modules (CIM) will continue use the string alternative method.
```
I am unsure what CIM is here

StevenBucher98 commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -94,9 +95,15 @@ public static class ApplicationInsightsTelemetry
         // private const string _psCoreTelemetryKey = "ee4b2115-d347-47b0-adb6-b19c2c763808"; // Production
         private const string _psCoreTelemetryKey = "d26a5ef4-d608-452c-a6b8-a4a55935f70d"; // V7 Preview 3
 
+        // In the event there is a problem in creating the node identifier file, use the default identifier.
```
Probably worth mentioning that this is typically within read only containers/env, unless it is unclear how many possible cases where there might be a problem creating node identifier.

StevenBucher98 replied:
Probably worth mentioning that this is typically within read only containers/env, unless it is unclear how many possible cases where there might be a problem creating node identifier.

StevenBucher98 commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -729,22 +756,30 @@ internal static void SendPSCoreStartupTelemetry(string mode)
+            // This is the payload for the parameter data which is sent as a metric.
+            var parameters = new Dictionary<string, double>();
+
             // The variable POWERSHELL_DISTRIBUTION_CHANNEL is set in our docker images.
```
```suggestion
            // The variable POWERSHELL_DISTRIBUTION_CHANNEL is set in our docker images and various Azure Services.
```
```suggestion
            // The variable POWERSHELL_DISTRIBUTION_CHANNEL is set in our docker images.
```

StevenBucher98 replied:
```suggestion
            // The variable POWERSHELL_DISTRIBUTION_CHANNEL is set in our docker images and various Azure Services.
```
```suggestion
            // The variable POWERSHELL_DISTRIBUTION_CHANNEL is set in our docker images.
```

JamesWTruher committed:
improve comment about why we have a default node identifier.

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -195,6 +194,53 @@ internal static int MaxNameLength()
+            ExecutionPolicy   = 0x00000010, // -ExecutionPolicy | -ex | -ep
+            File              = 0x00000020, // -File | -f
+            Help              = 0x00000040, // -Help, -?, /?
+            InputFormat       = 0x00000100, // -InputFormat | -inp | -if
```
Why did you skip 0x00000080?

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -195,6 +194,53 @@ internal static int MaxNameLength()
+            Version           = 0x00100000, // -Version | -v
+            WindowStyle       = 0x00200000, // -WindowStyle | -w
+            WorkingDirectory  = 0x00400000, // -WorkingDirectory | -wd
+            // Parameter values for ExecutionPolicy
```
I don't think this comment is really necessary

```suggestion
            // Enum values for ExecutionPolicy
```

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -641,6 +687,47 @@ internal static string NormalizeFilePath(string path)
+                return ParameterBitmap.EPUndefined;
+            }
+
+            if (MatchSwitch(_executionPolicy, "remotesigned", "remotesigned"))
```
Would this be better as a Dictionary that maps the string to the bitmap value?

JamesWTruher replied:
This allowed me to reuse code that we use already for parameter recognition.

daxian-dbw replied:
Given that you don't need to check on a substring match, maybe just use `_executionPolicy.Equals("remotesigned", StringComparison.OrdinalIgnoreCase)`?

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -763,27 +851,33 @@ private void ParseHelper(string[] args)
+                    ParametersUsed |= ParameterBitmap.NonInteractive;
                     _noInteractive = true;
                 }
                 else if (MatchSwitch(switchKey, "socketservermode", "so"))
```
Is there a reason not to cover the other server modes?

JamesWTruher replied:
they're covered in other parts of the telemetry (see ConsoleHost line:195)

SteveL-MSFT replied:
I saw that later, but you cover the ssh server which is also covered later

SteveL-MSFT commented on src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs:
```diff
@@ -619,7 +619,7 @@ private PSModuleInfo ImportModule_LocallyViaName_WithTelemetry(ImportModuleOptio
                 if (!foundModule.IsWindowsPowerShellCompatModule)
                 {
-                    ApplicationInsightsTelemetry.SendTelemetryMetric(TelemetryType.ModuleLoad, foundModule.Name);
+                    ApplicationInsightsTelemetry.SendModuleTelemetryMetric(TelemetryType.ModuleLoad, foundModule.Name, foundModule.Version.ToString());
```
Is there a case where Version is null like you cover on line 899 below?

SteveL-MSFT replied:
You use the null conditional operator in the other cases below, so it looks like an omission here?

JamesWTruher replied:
the PSModuleInfo structure does not have this member as nullable, it shouldn't need the protection

SteveL-MSFT commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -94,9 +95,16 @@ public static class ApplicationInsightsTelemetry
         private const string _anonymous = "anonymous";
 
+        // Use '0.0.0.0' as the string for an anonymous module version
+        private const string _anonymousVersion = "0.0.0.0";
```
Does this need all 4 segments of the version or can we save space with just "0"?

JamesWTruher replied:
I thought with an actual version it might be easier on the query side ([version]::new("0") will throw), i can make this "0.0" which can still be converted to a version type to reduce payload size.

SteveL-MSFT commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -813,35 +849,39 @@ private static bool TryCreateUniqueIdentifierAndFile(string telemetryFilePath, o
 
             // The directory may not exist, so attempt to create it
             // CreateDirectory will simply return the directory if exists
+            bool attemptFileCreation = true;
```
Seems like this should be called `telemetryDirectoryCreationFailed` being `false` by default

JamesWTruher replied:
i tried that the first time, but didn't like the readability logic in the if statement. I thought `if ( attemptFileCreation` read more easily than `if ( ! Failed`

SteveL-MSFT commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -813,35 +849,39 @@ private static bool TryCreateUniqueIdentifierAndFile(string telemetryFilePath, o
+                try
+                {
+                    id = Guid.NewGuid();
+                    File.WriteAllBytes(telemetryFilePath, id.ToByteArray());
```
Is there a case where you can create the directory but not the file?  Like hitting a filesystem quote situation.

JamesWTruher replied:
there could be, which is why i left this failure to send the saveuuid telemetry. If it does fail, then we'll set the node identifier to the default

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -849,9 +948,12 @@ private void ParseHelper(string[] args)
                 }
                 else if (MatchSwitch(switchKey, "windowstyle", "w"))
                 {
+                    ParametersUsed |= ParameterBitmap.WindowStyle;
```
Using this flag will fail on Unix, so maybe move this to the `#else` block?

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -879,6 +981,7 @@ private void ParseHelper(string[] args)
                 else if (MatchSwitch(switchKey, "file", "f"))
                 {
+                    ParametersUsed |= ParameterBitmap.File;
                     if (!ParseFile(args, ref i, noexitSeen))
```
I notice some parameter info is collected regardless of whether the arg parsing will fail, like this one, and some others is collected only if the arg parsing is successful, such as `settingsfile` or `workingdirectory`. It would be nice to be consistent.

daxian-dbw commented on src/System.Management.Automation/utils/Telemetry.cs:
```diff
@@ -813,35 +849,39 @@ private static bool TryCreateUniqueIdentifierAndFile(string telemetryFilePath, o
-            return false;
+            // all attempts to create an identifier have failed, so use the default node id.
+            id = _defaultNodeIdentifier;
+            return true;
```
Given this method always returns `true` now, you should change the method signature to just return the `Guid` identifier.
Also, for the call site of this method, no need to keep the `if` check, instead, it can be `return ryCreateUniqueIdentifierAndFile(uuidPath)`.

JamesWTruher committed:
Add more values to the parameter bitmap.

Address some code review issues:
- Changed the logic of how a uuid file is created, now if we can't create the file, we'll use a default identifier.
- Make the assignment of the parameter bitmap more consistent.
 - update some comments to elaborate a bit more.

JamesWTruher committed:
Turn off developer mode.

JamesWTruher committed:
Protect from a null Version property in the PSModuleInfo object.

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -641,6 +691,47 @@ internal static string NormalizeFilePath(string path)
+            {
+                return ParameterBitmap.EPRemoteSigned;
+            }
+            else if (String.Equals(_executionPolicy, "bypass", StringComparison.OrdinalIgnoreCase))
```
'else if' is not needed since we do an early return.

JamesWTruher replied:
turned the whole thing into a switch

JamesWTruher committed:
fix up code factor issue to use string instead of String.

SteveL-MSFT reviewed (approved):
LGTM

JamesWTruher committed:
Change get execution policy map to use a switch.

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -925,9 +1049,12 @@ private void ParseHelper(string[] args)
                 }
                 else if (MatchSwitch(switchKey, "sta", "sta"))
                 {
+                    ParametersUsed |= ParameterBitmap.STA;
```
This should be moved down before `_staMode = true;` to keep consistent with the rest -- collect info only if parsing is successful.

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -947,6 +1074,7 @@ private void ParseHelper(string[] args)
                 }
                 else if (MatchSwitch(switchKey, "mta", "mta"))
                 {
+                    ParametersUsed |= ParameterBitmap.MTA;
```
Same here, this line should be moved to `_staMode = false;` to keep consistent with the rest.

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -641,6 +691,51 @@ internal static string NormalizeFilePath(string path)
+            ParameterBitmap ExecutionPolicySetting = ParameterBitmap.EPUndefined;
+            switch (_executionPolicy)
+            {
+                case var b when string.Equals(_executionPolicy, "default", StringComparison.OrdinalIgnoreCase):
```
A minor comment: given that the var `b` is never used, it would be nicer to keep using `if`'s instead of a switch (the switch is actually also doing `if`'s in this case)
```
if (string.Equals(...))
{
    return xxx;
}

if (string.Equals(...))
{
    return xxx;
}

...
```

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -641,6 +691,51 @@ internal static string NormalizeFilePath(string path)
+            ParameterBitmap ExecutionPolicySetting = ParameterBitmap.EPUndefined;
+            switch (_executionPolicy)
+            {
+                case var b when string.Equals(_executionPolicy, "default", StringComparison.OrdinalIgnoreCase):
```
```suggestion
                case bool b when string.Equals(_executionPolicy, "default", StringComparison.OrdinalIgnoreCase):
```

JamesWTruher replied:
it's not a bool, it's actually a string - it's a new C# 7 thingy that surprisingly returns a string in this context.

PaulHigin replied:
What ever type it returned, it would be nice to make it explicit.  But not a blocker.

JamesWTruher replied:
and it's not used - I imagine (hope) that the compiler will just optimize it away since it's not used

PaulHigin reviewed (approved):
LGTM with switch statement suggestion.

JamesWTruher committed:
another codefactor nit.

JamesWTruher committed:
change execution policy parsing back to if/else.

JamesWTruher committed:
Add Az.Tools.Predictor to list of reported modules (requested by Steven).

daxian-dbw reviewed (approved):
LGTM

daxian-dbw committed:
Fix CodeFactor issues
